File: R/region.R
#===============================================================================
# region.R
#===============================================================================

# Imports ======================================================================

#' @import combinat
#' @import partitions




# Functions ====================================================================

#' @title Region of integration for bernoulli sums
#'
#' @param x number of successes
#' @param size vector giving the number of trials for each group.
#' @return a matrix with one column per sample, the rows of which are the
#'   coordinates of the region of integration.
region <- function(x, size) {
  if (x < 0 || x > sum(size)) stop("provided x is out of bounds")
  if (x == 0) return(matrix(0, nrow = 1, ncol = length(size)))
  if (x == sum(size)) return(matrix(size, nrow = 1))
  size_sorted <- sort(size, decreasing = TRUE)

  if (x == 1) {
    part <- matrix(c(1, rep(0, length(size) - 1)), ncol = 1)
  } else {
    if (x < length(size)) {
      print("x < length(size)")
      print(x)
      part <- parts(x)
      part <- rbind(
        part,
        matrix(0, nrow = length(size) - x, ncol = ncol(part))
      )
    } else {
      print("x >= length(size)")
      print(x)
      part <- parts(x)[1:length(size),, drop = FALSE]
    }
    part <- part[,
      apply(part, 2, function(col) all(col <= size_sorted) && (sum(col) >= x))
    ]
  }

  part <- do.call(
    cbind,
    lapply(
      apply(part, 2, function(col) {unique(permn(col))}),
      function(l) {do.call(cbind, l)}
    )
  )
  part <- part[,
    apply(
      part,
      2,
      function(col) {
        all(col <= size)
      }
    )
  ]
  t(part)
}

File: betabernsum/region.py
#===============================================================================
# region.R
#===============================================================================

# Imports ======================================================================

import itertools
from accelasc import accel_asc




# Functions ====================================================================

def region(x, size):
    """Region of integration for bernoulli sums

    Parameters
    ----------
    x : int
        number of successes
    size
        iterable of integers, giving the number of trials for each group
    
    Returns
    -------
    numpy array
        an array with one column per sample, the rows of which are the
        coordinates of the region of integration.
    """

    if x < 0 or x > sum(size):
        raise RuntimeError('provided x is out of bounds')
    if x == 0:
        yield (0,) * len(size)
        return
    if x == sum(size):
        yield tuple(size)
        return
    yield from itertools.chain.from_iterable(
        set(
            permutation for permutation in itertools.permutations(
                partition + [0] * (len(size) - len(partition))
            )
            if all((p <= s for p, s in zip(permutation, size)))
        )
        for partition in accel_asc(x) if len(partition) <= len(size)
    )
